### PR TITLE
Add an unsizing operation to {Arc,Rc}Box

### PR DESCRIPTION
**Pull request by HeroicKatora**

### CODE REVIEW

File: crates/rc-box/src/lib.rs [hunk 1]
```diff
@@ -634,6 +635,50 @@ then the data will be pinned in memory and unable to be moved."),
...
+
+let unique = ", stringify!($RcBox), "::new(|| 42u32);
+let unique:", stringify!($RcBox), r"<dyn Fn() -> u32> =
+    unique.unsize(Coercion::<_, dyn Fn() -> u32>::to_fn());
+
+let value = (*unique)();
+assert_eq!(value, 42);
+```
+
+Another common usage would be to create an `dyn Any`.
```
Comment (CAD97) [inline crates/rc-box/src/lib.rs:656]: nitpick: this should be either "create `dyn Any`" or "create a `dyn Any`", not "create an `dyn Any`" to fit the rest of pointer-utils' crates' documentation.

File: crates/rc-box/src/lib.rs [hunk 2]
```diff
@@ -634,6 +635,50 @@ then the data will be pinned in memory and unable to be moved."),
 
         #[cfg(feature = "std")]
         impl<T: ?Sized> UnwindSafe for $RcBox<T> where Box<T>: UnwindSafe {}
+
+        #[cfg(feature = "unsize")]
+        doc_comment! {
+            concat!("Unsizes a pointer using the `unsize` crate.
+# Usage
```
Comment (CAD97) [inline crates/rc-box/src/lib.rs:642]: nitpick: there should be another newline before the header here to match the style of the rest of the documentation.

File: crates/rc-box/src/lib.rs [hunk 3]
```diff
@@ -95,8 +95,6 @@ macro_rules! rc_box {
 
         // ~~~ $Rc<T> and Box<T> like inherent impls ~~~ //
 
-        // downcast is pretty useless without CoerceUnsized
```
Comment (CAD97) [inline crates/rc-box/src/lib.rs:98]: I'm glad to see this go 😄

**Review by CAD97 — APPROVED** [review]
Looks good to me! Honestly, `unsize` would fit right in with the pointer-utils family. (I've been thinking of de-mono-repoing and putting these crates in an org, if you'd want to collaborate. Just a thought, not an ask. I'm still not even sure I want to set up an org.) I just have a couple nits and one thought:

thought: should the other crates support explicit unsizing? Specifically, [`{A}RcBorrow<T>`](https://cad97.github.io/pointer-utils/rc_borrow/index.html) and [`Thin<Ptr<T>>`](https://cad97.github.io/pointer-utils/erasable/struct.Thin.html) seem like good candidates. (The latter would need to make sure that the coerced-to pointer is also erasable.)

I'm happy to merge as-is once you've thought about those impls. Either way, they should probably be added in a followup PR anyway.

bors: d+

**Comment by CAD97** [comment]
Oh, and I fixed the new warnings in #72, don't worry about them.

**Comment by HeroicKatora** [comment]
> thought: should the other crates support explicit unsizing? Specifically, {A}RcBorrow<T> and Thin<Ptr<T>> seem like good candidates. (The latter would need to make sure that the coerced-to pointer is also erasable.)

I'll see about a follow-up PR for those two crates, this sounds good to me as well. I wasn't sure about `Thin` initially but from the point of view of this being a true operation, not a builtin zero-cost coercion, and after viewing its implementation it does seem more reasonable.

**Comment by CAD97** [comment]
> I wasn't sure about `Thin` initially but from the point of view of this being a true operation, not a builtin zero-cost coercion, and after viewing its implementation it does seem more reasonable.

Yeah, erase/unerase is always just be a pointer cast for sized types, so coercing from `Thin<Box<Sized>>` to `Thin<Box<Unsized>>` should just be

- cast `ErasedPtr` to `*mut Erased`
- cast `*mut Erased` to `*mut Sized`
- cast `*mut Sized` to `Box<Sized>`
- coerce `Box<Sized>` to `Box<Unsized>`
- cast `Box<Unsized>` to `*mut Unsized`
- cast `*mut Unsized` to `*mut Erased`
- offset `*mut Erased` per `Unsized::erase`
- cast `*mut Erased` to `ErasedPtr`

all of which should be unlined.

**Pull request merged**